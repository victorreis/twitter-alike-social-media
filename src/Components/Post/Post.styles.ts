import styled, { CSSObject, DefaultTheme } from 'styled-components';

import { Typography } from '../Typography';

const getPostBorders = ({
  theme,
  compact,
}: {
  theme: DefaultTheme;
  compact: boolean;
}): CSSObject => {
  if (compact) {
    return {
      borderWidth: '1px',
      borderStyle: 'solid',
      borderColor: theme.colors.background.default.light,
      borderRadius: 16,
    };
  }

  return {
    borderBottomWidth: '1px',
    borderBottomStyle: 'solid',
    borderBottomColor: theme.colors.background.default.normal,
  };
};

const getPostContentButtonsContainerBorder = ({
  theme,
  hasChildren,
}: {
  theme: DefaultTheme;
  hasChildren: boolean;
}): CSSObject => {
  return {
    ...(!hasChildren && {
      borderTopWidth: '1px',
      borderTopStyle: 'solid',
      borderTopColor: theme.colors.background.default.normal,
    }),
  };
};

export const PostContainer = styled.div`
  display: flex;
  padding: 16px;
  column-gap: 1rem;

  ${getPostBorders}
`;

export const PostContent = styled.div`
  display: flex;
  flex-direction: column;
  column-gap: 1rem;
  flex: 1;
`;

export const PostContentHeader = styled.div`
  display: flex;
  column-gap: 1rem;
  align-items: center;
`;

export const PostContentTitle = styled(Typography)`
  margin: 0;
`;

export const PostText = styled(Typography)`
  flex: 1;
  margin-bottom: 16px;
`;

export const PostContentButtonsContainer = styled.div<{ hasChildren: boolean }>`
  display: flex;
  flex: 1;
  column-gap: 1rem;
  align-items: center;
  align-self: stretch;
  justify-content: flex-end;
  padding-top: 16px;

  ${getPostContentButtonsContainerBorder};
`;
