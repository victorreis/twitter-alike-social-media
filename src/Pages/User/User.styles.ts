import styled from 'styled-components';

export const UserContainer = styled.div`
  display: flex;
  flex-direction: column;
`;

export const UserTabBar = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
  align-items: flex-start;

  box-sizing: border-box;
  border-bottom: 1px solid
    ${({ theme }) => theme.colors.background.default.light};
`;
