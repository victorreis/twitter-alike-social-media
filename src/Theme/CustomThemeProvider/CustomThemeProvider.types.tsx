import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
