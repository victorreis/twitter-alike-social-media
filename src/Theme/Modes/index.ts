export { darkTheme } from './Dark.theme';
