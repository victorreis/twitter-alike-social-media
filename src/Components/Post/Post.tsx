import { useMemo, useRef, useState } from 'react';

import dayjs from 'dayjs';
import { nanoid } from 'nanoid';

import { EXTENDED_DATE_FORMAT } from '../../Config/Constants.config';
import { TestProps } from '../../Config/Tests/Test.types';
import { TypographyVariant } from '../../Theme/Types/Typographies.types';
import { Avatar } from '../Avatar';
import { Button, ButtonSize } from '../Button';
import { PostCreator } from '../PostCreator';
import { Typography } from '../Typography';
import {
  PostContainer,
  PostContent,
  PostContentButtonsContainer,
  PostContentHeader,
  PostContentTitle,
  PostText,
} from './Post.styles';
import { PostProps, DefaultPostProps } from './Post.types';

export const postDefaults: Required<DefaultPostProps> & Required<TestProps> = {
  testID: 'Post',
  compact: false,
};

export const Post: React.FC<PostProps> = (props): JSX.Element => {
  const {
    testID = postDefaults.testID,
    children,
    id,
    text,
    createdAt,
    createdBy,
    reposts,
    quotPosts,
    compact = postDefaults.compact,
    isAvatarClickable,
    onClickAvatar,
    onClickRepost,
    onClickQuotePost,
    style,
    ...others
  } = props;

  const { name, nickname, thumbnailUrl } = createdBy;
  const [showPostCreator, setShowPostCreator] = useState<boolean>();
  const quotePostCreatorRef = useRef<HTMLTextAreaElement>(null);

  const formattedDate = dayjs(createdAt).format(EXTENDED_DATE_FORMAT);

  const userNameVariant: TypographyVariant = compact ? 'h5' : 'h4';
  const userInfoVariant: TypographyVariant = compact
    ? 'subtitle3'
    : 'subtitle2';
  const textVariant: TypographyVariant = compact ? 'body2' : 'body1';
  const buttonSize: ButtonSize = compact ? 'MD' : 'LG';

  const handleClickRepost = () => {
    onClickRepost(id);
  };

  const handleSubmitQuotePost = (quotePostText: string) => {
    setShowPostCreator(() => false);
    onClickQuotePost(id, quotePostText);
  };

  const renderText = useMemo(
    () => () => {
      return text.split('\n').map((line) => (
        <span key={nanoid()}>
          {line}
          <br />
        </span>
      ));
    },
    [text]
  );

  const renderAvatar = () => {
    return (
      <Avatar
        isAvatarClickable={Boolean(isAvatarClickable)}
        name={name}
        nickname={nickname}
        onClickAvatar={onClickAvatar}
        size={compact ? 'SM' : 'MD'}
        thumbnailUrl={thumbnailUrl}
      />
    );
  };

  const handleShowQuotePostCreator = () => {
    setShowPostCreator(() => true);
    if (quotePostCreatorRef.current) {
      quotePostCreatorRef.current.focus();
    }
  };

  const renderQuotePostCreator = () => {
    return (
      showPostCreator && (
        <PostCreator
          ref={quotePostCreatorRef}
          isAvatarClickable={isAvatarClickable}
          name={name}
          nickname={nickname}
          onClickAvatar={onClickAvatar}
          onSubmit={handleSubmitQuotePost}
          thumbnailUrl={thumbnailUrl}
        />
      )
    );
  };

  return (
    <PostContainer
      key={id}
      compact={compact}
      data-testid={testID}
      style={style}
      {...others}
    >
      {!compact && renderAvatar()}

      <PostContent>
        <PostContentHeader>
          {compact && renderAvatar()}

          <PostContentTitle variant={userNameVariant}>{name}</PostContentTitle>
          <PostContentTitle variant={userInfoVariant}>
            @{nickname}
          </PostContentTitle>
          <Typography variant={userInfoVariant}>- {formattedDate}</Typography>
        </PostContentHeader>

        <PostText data-testid={`${testID}_text`} variant={textVariant}>
          {renderText()}
        </PostText>

        {children}

        <PostContentButtonsContainer hasChildren={Boolean(children)}>
          <Button onClick={handleClickRepost} size={buttonSize}>
            Repost
          </Button>
          <Typography variant={textVariant}>{reposts}</Typography>
          <Button
            disabled={showPostCreator}
            onClick={handleShowQuotePostCreator}
            size={buttonSize}
          >
            Quote post
          </Button>

          <Typography variant={textVariant}>{quotPosts}</Typography>
        </PostContentButtonsContainer>

        {renderQuotePostCreator()}
      </PostContent>
    </PostContainer>
  );
};
