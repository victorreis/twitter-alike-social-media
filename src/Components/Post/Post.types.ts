import { TestProps } from '../../Config/Tests/Test.types';
import { PostType } from '../../Models/Post.types';
import { AvatarClickProps } from '../Avatar';

export interface RequiredPostProps {
  /**
   * Callback function that is called when the Repost button is clicked.
   */
  onClickRepost: (originalPostId: string) => void;

  /**
   * Callback function that is called when the QuotePost button is clicked.
   */
  onClickQuotePost: (originalPostId: string, quotePostText: string) => void;
}

export interface DefaultPostProps {
  /**
   * Controls when the Post is compact or not.
   */
  compact?: boolean;
}

export interface OptionalPostProps {}

export type PostProps = RequiredPostProps &
  DefaultPostProps &
  OptionalPostProps &
  TestProps &
  PostType &
  AvatarClickProps &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;
