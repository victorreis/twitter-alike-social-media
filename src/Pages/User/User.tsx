import { useState, useEffect } from 'react';
import { toast } from 'react-toastify';

import { PageContainer } from '../../App.styles';
import { PostCreator } from '../../Components/PostCreator';
import { Tab } from '../../Components/Tab';
import { UserDetails } from '../../Components/UserDetails';
import {
  REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY,
  UNNAVAILABLE_USER,
} from '../../Config/ErrorMessages.config';
import {
  FOLLOW_CREATED,
  FOLLOW_DELETED,
  POST_CREATED,
} from '../../Config/SuccessMessages.config';
import { TestProps } from '../../Config/Tests/Test.types';
import { TOAST_PARAMS } from '../../Config/Toast.config';
import { useLoggedUser } from '../../Hooks/LoggedUser';
import { useRenderPosts } from '../../Hooks/RenderPosts';
import { useShowUserPage } from '../../Hooks/ShowUserPage';
import { followerFollowedCreatorService } from '../../Services/FollowerFollowedCreator';
import { followerFollowedRetrieverService } from '../../Services/FollowerFollowedRetriever';
import { postCreatorService } from '../../Services/PostCreator';
import { UserTabBar } from './User.styles';

export const userDefaults: Required<TestProps> = {
  testID: 'User',
};

export const User: React.FC = (): JSX.Element | null => {
  const [isFollowing, setIsFollowing] = useState<boolean | undefined>(
    undefined
  );
  const { userFromUserPage, isAvatarClickable, handleShowUserPage } =
    useShowUserPage();
  const { loggedUser } = useLoggedUser();
  const { renderPosts, updateLoadedPosts } = useRenderPosts({
    userId: userFromUserPage?.id,
  });

  useEffect(() => {
    if (
      loggedUser &&
      userFromUserPage &&
      loggedUser?.id !== userFromUserPage?.id
    ) {
      setIsFollowing(() =>
        followerFollowedRetrieverService.isFollowerFollowed(
          loggedUser.id,
          userFromUserPage.id
        )
      );
    }
  }, [loggedUser, userFromUserPage, userFromUserPage?.id]);

  if (!userFromUserPage) {
    return null;
  }

  const handleFollow = () => {
    if (loggedUser) {
      try {
        followerFollowedCreatorService.createRelationship(
          loggedUser.id,
          userFromUserPage.id
        );
        toast.success(
          FOLLOW_CREATED(userFromUserPage.nickname),
          TOAST_PARAMS['SUCCESS']
        );
        setIsFollowing((prevState) => !prevState);
        updateLoadedPosts();
      } catch (e) {
        toast.error(UNNAVAILABLE_USER, TOAST_PARAMS['ERROR']);
      }
    }
  };

  const handleUnfollow = () => {
    if (loggedUser) {
      try {
        followerFollowedCreatorService.deleteRelationship(
          loggedUser.id,
          userFromUserPage.id
        );
        toast.success(
          FOLLOW_DELETED(userFromUserPage.nickname),
          TOAST_PARAMS['SUCCESS']
        );
        setIsFollowing((prevState) => !prevState);
        updateLoadedPosts();
      } catch (e) {
        toast.error(UNNAVAILABLE_USER, TOAST_PARAMS['ERROR']);
      }
    }
  };

  const handleSubmitPost = (text: string) => {
    if (loggedUser) {
      try {
        postCreatorService.createPost(text, loggedUser.id);
        toast.success(POST_CREATED, TOAST_PARAMS['SUCCESS']);
        updateLoadedPosts();
      } catch (e) {
        toast.error(
          REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY,
          TOAST_PARAMS['ERROR']
        );
      }
    }
  };

  const renderPostCreator = () => {
    const { id, name, thumbnailUrl } = userFromUserPage;
    if (loggedUser?.id !== id) {
      return null;
    }

    return (
      <PostCreator
        isAvatarClickable={isAvatarClickable}
        name={name}
        nickname={String(loggedUser?.nickname)}
        onClickAvatar={handleShowUserPage}
        onSubmit={handleSubmitPost}
        thumbnailUrl={thumbnailUrl}
      />
    );
  };

  return (
    <PageContainer data-testid={userDefaults.testID}>
      <UserDetails
        {...userFromUserPage}
        isFollowing={isFollowing}
        onFollow={handleFollow}
        onUnfollow={handleUnfollow}
      />

      {renderPostCreator()}

      <UserTabBar>
        <Tab active text="POSTS" />
      </UserTabBar>

      {renderPosts()}
    </PageContainer>
  );
};
