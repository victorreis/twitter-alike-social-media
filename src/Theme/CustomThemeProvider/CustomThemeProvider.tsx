/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
