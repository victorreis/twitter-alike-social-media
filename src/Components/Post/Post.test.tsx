import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { PostType } from '../../Post.types';
import { themes } from '../../Theme/CustomThemeProvider';
import { UserType } from '../../User.types';
import { Post, postDefaults } from './Post';
import { PostProps } from './Post.types';

describe('post component tests', () => {
  const userTeste: UserType = {
    id: 'user123',
    name: 'Victor Reis',
    nickname: 'victor.reis',
    url: '...',
    thumbnailUrl: 'https://i.pravatar.cc/300',
    createdAt: new Date(),
    following: 333,
    followers: 777,
  };
  const post: PostType = {
    id: 'post123',
    text: `first line\nsecond line\nthird line`,
    createdAt: new Date(),
    createdBy: userTeste,
    reposts: 9,
    quotPosts: 99,
  };
  const newCompact = true;

  const requiredProps: PostProps = {
    ...post,
  };

  const setup = (props?: PostProps) => {
    const renderRTR = () =>
      renderRTRCreator<PostProps>(Post, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<PostProps>(Post, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(postDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text`, () => {
      expect.assertions(4);
      setup().renderJestDom();
      const lines = post.text.split('\n');
      const firstLine = screen.getByTestId(`${postDefaults.testID}_text`);

      expect(firstLine).toBeInTheDocument();
      expect(firstLine).toHaveTextContent(String(lines[0]));
      expect(firstLine).toHaveTextContent(String(lines[1]));
      expect(firstLine).toHaveTextContent(String(lines[2]));
    });

    it(`should render '${String(
      postDefaults.compact
    )}' as the default 'compact'`, () => {
      expect.assertions(1);
      const instance = setup().renderRTR().root;
      const element = instance.findByProps({
        compact: postDefaults.compact,
      });

      expect(element).toBeTruthy();
    });

    it(`should override the default 'compact' when it is passed as prop`, () => {
      expect.assertions(1);
      const instance = setup({
        ...requiredProps,
        compact: newCompact,
      }).renderRTR().root;
      const element = instance.findByProps({ compact: newCompact });

      expect(element).toBeTruthy();
    });
  });

  describe('style tests', () => {
    it(`should have styled the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(postDefaults.testID);

      expect(container).toHaveStyle({
        borderBottomWidth: '1px',
        borderBottomStyle: 'solid',
        borderBottomColor: themes.default.colors.background.default.light,
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
