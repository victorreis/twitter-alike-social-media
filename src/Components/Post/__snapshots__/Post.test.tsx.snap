// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post component tests snapshot tests should render correctly 1`] = `
<div
  className="PostContainer-twitter-alike-social-media__sc-zy8jam-0 gsEoCj"
  data-testid="Post"
  type="Post"
>
  <div
    className="AvatarContainer-twitter-alike-social-media__sc-ne4tqs-0 czdfxE"
    data-testid="Avatar"
    onClick={[Function]}
    size="MD"
  >
    <img
      alt="Victor Reis's avatar photo."
      className="AvatarImage-twitter-alike-social-media__sc-ne4tqs-1 dnbueO"
      data-testid="Avatar_image"
      onError={[Function]}
      size="MD"
      src="https://i.pravatar.cc/300"
      title="Victor Reis's avatar photo."
    />
  </div>
  <div
    className="PostContent-twitter-alike-social-media__sc-zy8jam-1 kvzPWr"
  >
    <div
      className="PostContentHeader-twitter-alike-social-media__sc-zy8jam-2 ilvDfQ"
    >
      <h4
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 dwIaAW PostContentTitle-twitter-alike-social-media__sc-zy8jam-3 bFlsZm"
        data-testid="Typography"
      >
        Victor Reis
      </h4>
      <h4
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 cCEHbN PostContentTitle-twitter-alike-social-media__sc-zy8jam-3 bFlsZm"
        data-testid="Typography"
      >
        @
        victor.reis
      </h4>
      <h4
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 cCEHbN"
        data-testid="Typography"
      >
        - 
        January 1, 2011
      </h4>
    </div>
    <span
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT PostText-twitter-alike-social-media__sc-zy8jam-4 kKAkLD"
      data-testid="Post_text"
    >
      <span>
        first line
        <br />
      </span>
      <span>
        second line
        <br />
      </span>
      <span>
        third line
        <br />
      </span>
    </span>
    <div
      className="PostContentButtonsContainer-twitter-alike-social-media__sc-zy8jam-5 gsSVXy"
    >
      <button
        className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 cpQvHN"
        data-testid="Button"
        disabled={false}
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        size="LG"
        type="button"
      >
        Repost
      </button>
      <span
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
        data-testid="Typography"
      >
        9
      </span>
      <button
        className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 cpQvHN"
        data-testid="Button"
        disabled={false}
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        size="LG"
        type="button"
      >
        Quote post
      </button>
      <span
        className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
        data-testid="Typography"
      >
        99
      </span>
    </div>
  </div>
</div>
`;
